### ISSUE TITLE
Word with max= parses incorrectly

### ISSUE THREAD

**Issue by panda-34** [pyparsing/pyparsing#409]
`Word('AB', 'C', max=2).parse_string('A', parse_all=True)` gives ParsingException whereas `Word('AB', 'C', max=3).parse_string('A', parse_all=True)` works correctly. The error didn't happen in v 2.4

**Comment by ptmcg** [comment]
Looking at this - thanks.

Using this code to debug (exception when max=1 is expected):

``` python
for mx in range(1, 5):
    expr = Word('AB', 'C', max=mx)
    expr.runTests(f"""\
        # max={mx}
        A
        B
        BC
        """)
    print()

```

**Comment by ptmcg** [comment]
Fixed, simplified code, AND added optimization when 'exact' arg is used. Nice catch, thanks!

**Comment by ptmcg** [comment]
Released in 3.1.0a1
